arch/arm/src/armv7-a/Toolchain.defs

```diff
@@ -56,21 +56,51 @@ endif
 #
 
 ifeq ($(CONFIG_ARM_THUMB),y)
-ARCHCPUFLAGS += -mthumb
-endif
-
-ifeq ($(CONFIG_ARCH_FPU),y)
-ARCHCPUFLAGS += -mfloat-abi=hard
+  ARCHCPUFLAGS += -mthumb
 endif
 
 ifeq ($(CONFIG_ARCH_CORTEXA5),y)
-ARCHCPUFLAGS += -mcpu=cortex-a5
+  ARCHCPUFLAGS += -mcpu=cortex-a5
 else ifeq ($(CONFIG_ARCH_CORTEXA7),y)
-ARCHCPUFLAGS += -mcpu=cortex-a7
+  ARCHCPUFLAGS += -mcpu=cortex-a7
 else ifeq ($(CONFIG_ARCH_CORTEXA8),y)
-ARCHCPUFLAGS += -mcpu=cortex-a8
+  ARCHCPUFLAGS += -mcpu=cortex-a8
 else ifeq ($(CONFIG_ARCH_CORTEXA9),y)
-ARCHCPUFLAGS += -mcpu=cortex-a9
+  ARCHCPUFLAGS += -mcpu=cortex-a9
+endif
+
+ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
+  ARCHFPUD32FLAGS = -d16
+endif
+
+ifeq ($(CONFIG_ARM_FP_NEON),y)
+  ARCHNEONFLAGS = neon-
+  ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
+    $(error "Number of FPU is alwasys 32 with NEON enabled, please check Kconfig of your SoC.")
+  endif
+endif
+
+ifeq ($(CONFIG_ARCH_FPU),y)
+  ifeq ($(CONFIG_ARM_FPU_ABI_SOFT),y)
+    ARCHCPUFLAGS += -mfloat-abi=softfp
+  else
+    ARCHCPUFLAGS += -mfloat-abi=hard
+  endif
+  ifeq ($(CONFIG_ARCH_CORTEXA7),y)
+    ifneq ($(CONFIG_ARM_FP_NEON),y)
+      ARCHFPUVERSION = vfpv4
+    endif
+  else
+    ARCHFPUVERSION = vfpv3
```

Comment:
```suggestion
    ifeq ($(CONFIG_ARM_FP_NEON),y)
      ARCHFPUVERSION =
      ifeq ($(CONFIG_ARCH_CORTEXA8),y)
        ARCHNEONFLAGS = neon
      endif
    else
      ARCHFPUVERSION = vfpv3
    endif
```

Reply:
Ok. Hold on. I just found info that `-mfpu=neon (alias for neon-vfpv3)`. I will check if `-mfpu=neon-vfpv3` works and get back here

Reply:
Ok. Really GCC docs states that `Note that ‘neon’ is an alias for ‘neon-vfpv3’ and ‘vfp’ is an alias for ‘vfpv2’.`

Reply:
@no1wudi we can get back previous logic with less ifs. Sorry, my comment was faster than my searching in GCC docs.

Reply:
OK

Reply:
Looking into the table that you provided I think the implementation does not match.
We should have something like:
remove `ARCHNEONFLAGS` completely together with
```
ifeq ($(CONFIG_ARM_NEON),y)
  ARCHNEONFLAGS = neon-
  ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
    $(error "Number of FPU is alwasys 32 with NEON enabled, please check Kconfig of your SoC.")
  endif
endif
```
change
```
ifeq ($(CONFIG_ARCH_CORTEXA7),y)
  ifeq ($(CONFIG_ARM_NEON),y)
    ARCHFPUVERSION = neon-vfpv4
  else
    ARCHFPUVERSION = vfpv4
  endif
else
  ifeq ($(CONFIG_ARM_NEON),y)
    ARCHFPUVERSION = neon
  else
    ARCHFPUVERSION = vfpv3
  endif
endif

ARCHCPUFLAGS += -mfpu=$(ARCHFPUVERSION)$(ARCHFPUD32FLAGS)
```

Reply:
```
  ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
    $(error "Number of FPU is alwasys 32 with NEON enabled, please check Kconfig of your SoC.")
  endif
```
I think this should be handled by `Kconfig`

Reply:
And possible add `ARCHFPUSUFIX` to be
```
ifeq ($(CONFIG_ARCH_CORTEXA7),y)
  ARCHFPUSUFIX = -fp16
else ifeq ($(CONFIG_ARCH_CORTEXA7),y)
  ARCHFPUSUFIX = -fp16
endif
```
and
```
ARCHCPUFLAGS += -mfpu=$(ARCHFPUVERSION)$(ARCHFPUD32FLAGS)$(ARCHFPUSUFIX)
```

Reply:
Yes, but not all neon extension support FP16, so we should add a new option ARM_HAVE_NEON_FP16 for it

Reply:
> ```
>   ifneq ($(CONFIG_ARM_HAVE_FPU_D32),y)
>     $(error "Number of FPU is alwasys 32 with NEON enabled, please check Kconfig of your SoC.")
>   endif
> ```
> 
> I think this should be handled by `Kconfig`

Removed since NEON depend on CONFIG_ARM_HAVE_FPU_D32